File: advent_of_code_2017/day3/manhattandistance.py
def dimension(n):
    m = 0
    while n > (2*m+1)**2:
        m += 1
    return 2*m+1

def pos(n):
    dim = dimension(n)
    last_line = dim**2 - (dim-1)
    first_column = (dim - 1)**2 + 1
    first_line = (dim - 2)**2 + (dim - 1)
    last_column = (dim - 2)**2 + (dim - 1)
    # print(dict(n=n, dim=dim, last_line=last_line, first_column=first_column, first_line=first_line, last_column=last_column))
    if n >= last_line:
        return (dim-1, n - last_line)
    elif n >= first_column:
        return (n - first_column, 0)
    elif n >= first_line:
        return (0, n - first_line)
    return (last_column - n, dim - 1)

def distance(n):
    dim = dimension(n)
    pos1 = (dim-1)/2
    (a,b) = pos(n)
    return abs(a-pos1) + abs(b-pos1)

print(pos(22))
print(pos(17))
print(pos(4))
print(pos(6))
print(pos(8))
print(pos(2))
print(pos(368078))
print(distance(1))
print(distance(12))
print(distance(23))
print(distance(1024))
print(distance(368078))

File: advent_of_code_2017/day3/manhattandistance.hs
main :: IO ()
main = do
  putStrLn (show (dimension 24))
  putStrLn ("part1: " ++ (show (distance 368078)))

elm :: (Int, Int) -> Int
elm (row, col) = do
  let max_rowcol = max (abs row) (abs col)
  let dim = max_rowcol * 2 + 1
  if row == -max_rowcol
    then dim ^ 2 - (max_rowcol - col)
    else if col == -max_rowcol
           then dim ^ 2 - dim + 1 - (max_rowcol + row)
           else if row == max_rowcol
                  then dim ^ 2 - 2 * dim + 2 - (max_rowcol + col)
                  else dim ^ 2 - 3 * dim + 3 - (max_rowcol - row)

dimension :: Int -> Int
dimension n = head ([2 * m + 1 | m <- [1 ..], n <= (2 * m + 1) ^ 2])

pos :: Int -> (Int, Int)
pos idx = do
  let dim = dimension idx
  let max_rowcol = div (dim - 1) 2
  let bottom_left = dim ^ 2 - dim + 1
  let top_left = dim ^ 2 - 2 * dim + 2
  let top_right = dim ^ 2 - 3 * dim + 3
  if idx >= bottom_left
    then (-max_rowcol, max_rowcol - (dim ^ 2 - idx))
    else if idx >= top_left
           then (-max_rowcol + (bottom_left - idx), -max_rowcol)
           else if idx >= top_right
                  then (max_rowcol, -max_rowcol + (top_left - idx))
                  else (max_rowcol - (top_right - idx), max_rowcol)

distance :: Int -> Int
distance n = do
  let (a, b) = pos n
  abs a + abs b
